src/app/shared/services/media-state.service.ts

```diff
@@ -1,33 +1,81 @@
 import { BehaviorSubject, lastValueFrom, Observable } from 'rxjs'
-import { Injectable } from '@angular/core'
+import { Injectable, OnDestroy } from '@angular/core'
 import { MediaService } from './media.service'
 import { Media } from './media'
+import { EventSubscription } from '../../event-system/abstractions/event-observer.service'
+import { MediaEventObserver } from '../../core/services/media-event-observer.service'
+import { MediaCreatedEvent, MediaDeletedEvent } from '../../core/models/media-event'
+import { Logger } from '../../core/abstractions/logger.service'
 
 @Injectable()
-export class MediaStateService {
-  private readonly mediaSubject: Map<string, BehaviorSubject<Media>> = new Map()
+export class MediaStateService implements OnDestroy {
+  private readonly mediaSubjects: Map<string, BehaviorSubject<Media | undefined>> = new Map()
+  private readonly subscriptions: EventSubscription[] = []
+  private readonly logger: Logger
 
-  constructor(private readonly mediaService: MediaService) {}
+  constructor(
+    private readonly mediaService: MediaService,
+    private readonly mediaEventObserver: MediaEventObserver,
+    logger: Logger
+  ) {
+    this.logger = logger.tag('MediaStateService')
+    lastValueFrom(this.mediaService.getAllMedia())
+      .then(mediaAssets => mediaAssets.forEach(media => this.updateMedia(media)))
+      .catch(error => this.logger.data(error).error(`Failed while fetching all media from server`))
+    this.subscribeToEvents()
+  }
+
+  private updateMedia(media: Media): void {
+    const mediaSubject: BehaviorSubject<Media | undefined> | undefined = this.mediaSubjects.get(media.sourceName)
+    if (mediaSubject) {
+      mediaSubject.next(media)
+      return
+    }
+    this.mediaSubjects.set(this.getMediaSubjectKey(media.sourceName), new BehaviorSubject<Media | undefined>(media))
+  }
+
+  private getMediaSubjectKey(sourceName: string): string {
+    return sourceName.toUpperCase()
+  }
```

Comment:
Here you uppercase the key of the `mediaSubjects` but further down on line 52 you just set the `media.sourceName` as the key directly.
That seems like it will lead to multiple mediaSubjects for the same Media.

Reply:
Good catch, using it on line 34 was redundant since those values are the uppercased ones that we want to match when we create new subjects with undefined values from the piece metadata!